text/0000-iri-v181-transaction-and-bundle/0000-iri-v181-transaction-and-bundle.md

```diff
@@ -0,0 +1,980 @@
++ `IncomingBundleBuilder`, a builder pattern to create `Bundle` from a set of already created `Transaction`s, such as
+  those coming in over the wire;
++ `OutgoingBundleBuilder` and `SealedBundleBuilder`, two builder patterns working in tandem to create a `Bundle` from
+  `TransactionBuilder`s, i.e. transactions that are not yet fully constructed and will be finalized during the
+  construction process of the entire `Bundle`.
+
+In this proposal we keep the representation of the various types relatively simple and flat. We assume that the user is
+responsible for pushing the appropriate number of transaction drafts into a bundle builder and don't encode this on the
+type level. During the construction phase we check that all transactions are consistent and adhere to IOTA convention,
+but we don't, for example:
+
++ attempt to represent different the different types of transactions on the type level;
++ allow the bundle builders to push extra transactions into their collection automatically to encode a signature of
+  a certain size.
+
+The present proposal is meant as a building block for client facing code to provide higher level types that provide
+more convenient and implicit ways of constructing transactions and bundles.
+
+A transaction is 8019 trits large in total, out of which 6551 trits are available to store a payload. The payload is
+defined to either hold a signature fragment or a message fragment. Since it has a limited size, a user often needs more
+than one transaction to execute an operation. For example, signatures with [security level 2 or 3](Security levels)
+don't fit in a single transaction, and user-provided messages may exceed the maximum payload capacity so they need to be
+fragmented across multiple transactions. Moreover, because the total amount of tokens stored in the ledger has to stay
+constant, *input transactions* (which are called thus because an address is *put into* a transfer as a source of tokens,
+thus removing them from the address) have to always be matched with *output transactions* such that their total value is
+zero. For these reasons, transactions have to be processed as a whole in groups called bundles. A bundle is an atomic
+operation in the sense that either all or none of its transactions are accepted by the network. Even single transactions
+are propagated through the network within a bundle.
+
+The `Transaction` and `Bundle` types are intended to be final and immutable. This contract is enforced by not exposing
+any methods that allow manipulating their fields directly. `Transaction`s are intended to be constructed through
+the `TransactionBuilder` builder pattern, while `Bundle`s can be created either via the `IncomingBundleBuilder` or
+`OutgoingBundleBuilder`. The `IncomingBundleBuilder` takes in existing `Transaction`s, compiles them, and verifies
+their veracity before constructing a `Bundle` type. The `OutgoingBundleBuilder` takes in `TransactionBuilder`s,
```

Comment:
i'm not sure 'veracity' is the correct term here. Sounds more like a metaphysical term. What about 'validity'?

Reply:
Fixed. That formulation was awful. :)